### PR TITLE
feat(behavior_path_planner): add new turn signal algorithm

### PR DESCRIPTION
**Pull request by purewater0901**
## Description
Currently, the turn signal in Autoware does not work properly in some complicated situations. In this PR, I restructured the turn signal algorithm and make it work under super complicated situations. I also update the turn signal algorithm to follow the Japanese Road Traffic Act. 

[turn_signal_documentation(under_construction).pdf](https://github.com/autowarefoundation/autoware.universe/files/9661964/default.pdf)

## Tests performed
PSim and Unit Test

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Comment by purewater0901** [comment]
Internal test results 803/825

File: planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp [hunk 1]
```diff
@@ -2546,25 +2546,44 @@ TurnSignalInfo AvoidanceModule::calcTurnSignalInfo(const ShiftedPath & path) con
...
+  TurnSignalInfo turn_signal_info{};
+
+  if (std::abs(getRelativeLength(front_shift_point)) < 0.1) {
+    return turn_signal_info;
   }
 
-  // calc distance from ego to latest_shift_point end point.
-  if (turn_info.second >= 0.0) {
-    turn_signal.signal_distance = turn_info.second;
+  const auto signal_prepair_distance = std::max(getEgoSpeed() * 3.0, 10.0);
```
Comment (rej55) [inline planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp:2562]:
```suggestion
  const auto signal_prepare_distance = std::max(getEgoSpeed() * 3.0, 10.0);
```

File: planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp [hunk 2]
```diff
@@ -2546,25 +2546,44 @@ TurnSignalInfo AvoidanceModule::calcTurnSignalInfo(const ShiftedPath & path) con
...
+  const auto ego_to_shift_start =
+    calcSignedArcLength(path.path.points, getEgoPosition(), front_shift_point.start.position);
+
+  if (signal_prepair_distance < ego_to_shift_start) {
```
Comment (rej55) [inline planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp:2566]:
```suggestion
  if (signal_prepare_distance < ego_to_shift_start) {
```

**Review by rej55 — COMMENTED** [review]
Thank you for your contribution.
I add some suggestions to fix typo.
Please check.

File: planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp (hunk 1 above)
Comment (satoshi-ota) [reply planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp]: :pray:

**Review by rej55 — APPROVED** [review]
LGTM

**Pull request merged**